### PR TITLE
Add support for varchar to timestamp coercion in hive tables

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

Allows coerce a varchar type to timestamp in hive tables. 

## Additional context and related issues
Depends on https://github.com/trinodb/trino/pull/18004

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Add support for varchar to timestamp coercion in hive tables 
```

### CODE REVIEW

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 1]
```diff
@@ -129,11 +187,38 @@ public Object[][] timestampValuesProvider()
         };
     }
 
+    @DataProvider
+    public Object[][] invalidValue()
+    {
+        return new Object[][] {
+                {"Invalid timestamp"},
+                {"2022"},
+                {"2001-04-01T00:13:42.000"},
```
Comment (atanasenko) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:196]: Can you comment on why are those values invalid as timestamps?

**Review by atanasenko — APPROVED** [review]
PR description says "varchar type to string"

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 2]
```diff
@@ -102,6 +120,46 @@ public void testHistoricalLongTimestampToVarchar()
...
+    {
+        LocalDateTime localDateTime = LocalDateTime.parse("1899-12-31T23:59:59.999999");
+        SqlTimestamp timestamp = SqlTimestamp.fromSeconds(TIMESTAMP_MICROS.getPrecision(), localDateTime.toEpochSecond(UTC), localDateTime.get(NANO_OF_SECOND));
+        assertThatThrownBy(() -> assertVarcharToShortTimestampCoercions(
+                createUnboundedVarcharType(),
+                Slices.utf8Slice("1899-12-31 23:59:59.999999"),
+                TIMESTAMP_MICROS,
+                timestamp.getEpochMicros()))
+                .isInstanceOf(TrinoException.class)
+                .hasMessageContaining("Coercion on historical dates is not supported");
```
Comment (krvikash) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:146]: Maybe we can reformat this

```suggestion
        assertThatThrownBy(() ->
                assertVarcharToShortTimestampCoercions(
                        createUnboundedVarcharType(), 
                        Slices.utf8Slice("1899-12-31 23:59:59.999999"), 
                        TIMESTAMP_MICROS, 
                        timestamp.getEpochMicros()))
                .isInstanceOf(TrinoException.class)
                .hasMessageContaining("Coercion on historical dates is not supported");
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 3]
```diff
@@ -102,6 +120,46 @@ public void testHistoricalLongTimestampToVarchar()
...
+    {
+        LocalDateTime localDateTime = LocalDateTime.parse("1899-12-31T23:59:59.999999");
+        SqlTimestamp timestamp = SqlTimestamp.fromSeconds(TIMESTAMP_PICOS.getPrecision(), localDateTime.toEpochSecond(UTC), localDateTime.get(NANO_OF_SECOND));
+        assertThatThrownBy(() -> assertVarcharToShortTimestampCoercions(
+                createUnboundedVarcharType(),
+                Slices.utf8Slice("1899-12-31 23:59:59.999999"),
+                TIMESTAMP_PICOS,
+                timestamp.getEpochMicros()))
+                .isInstanceOf(TrinoException.class)
+                .hasMessageContaining("Coercion on historical dates is not supported");
```
Comment (krvikash) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:160]: Same here

```suggestion
        assertThatThrownBy(() ->
                assertVarcharToShortTimestampCoercions(
                        createUnboundedVarcharType(), 
                        Slices.utf8Slice("1899-12-31 23:59:59.999999"), 
                        TIMESTAMP_PICOS, 
                        timestamp.getEpochMicros()))
                .isInstanceOf(TrinoException.class)
                .hasMessageContaining("Coercion on historical dates is not supported");
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java [hunk 4]
```diff
@@ -129,11 +187,38 @@ public Object[][] timestampValuesProvider()
...
 
+    @DataProvider
+    public Object[][] invalidValue()
+    {
+        return new Object[][] {
+                {"Invalid timestamp"},
+                {"2022"},
+                {"2001-04-01T00:13:42.000"},
+                {"2001-14-01 00:13:42.000"},
+                {"2001-01-32 00:13:42.000"},
```
Comment (krvikash) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java:198]: Can we include one entry where time is invalid? Eg: `2001-04-01 23:59:60`

**Review by krvikash — COMMENTED** [review]
LGTM

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 5]
```diff
@@ -83,4 +89,64 @@ protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int pos
...
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            try {
+                Slice value = fromType.getSlice(block, position);
+                LocalDateTime dateTime = LOCAL_DATE_TIME.parse(value.toStringUtf8(), LocalDateTime::from);
+                long epochSecond = dateTime.toEpochSecond(UTC);
+                if (epochSecond < START_OF_MODERN_ERA_SECONDS) {
+                    throw new TrinoException(HIVE_INVALID_TIMESTAMP_COERCION, "Coercion on historical dates is not supported");
```
Comment (vlad-lyutenko) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:109]: Why we throw exception here, we are not able to adjust date using GregorianCalendar and `month -1 ` ?

Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:109]: Hive supports different way of storing timestamp across various version (2 and 3)  and various types (ORC/Parquet/RC) for some it uses `java.sql.Timestamp` and other it uses a different way - so we are restricting them

Comment (vlad-lyutenko) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:109]: So we just cannot predict what is the way of storing such date in underlying datasource (is it for example shifted for -1 month or not), so we just deny such cases.

Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:109]: It is not storing but actually the way it coerces from String to Timestamp

**Review by vlad-lyutenko — APPROVED** [review]

**Comment by Praveen2112** [comment]
@krvikash AC

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 6]
```diff
@@ -83,4 +89,64 @@ protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int pos
                             toType));
         }
     }
+
+    public static class VarcharToShortTimestampCoercer
+            extends TypeCoercer<VarcharType, TimestampType>
+    {
+        public VarcharToShortTimestampCoercer(VarcharType fromType, TimestampType toType)
+        {
+            super(fromType, toType);
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:98]: Might be useful check argument that TimestampType is short, given that we do `9 - toType.getPrecision()`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java [hunk 7]
```diff
@@ -83,4 +89,64 @@ protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int pos
...
+            }
+        }
+    }
+
+    public static class VarcharToLongTimestampCoercer
+            extends TypeCoercer<VarcharType, TimestampType>
+    {
+        public VarcharToLongTimestampCoercer(VarcharType fromType, TimestampType toType)
+        {
+            super(fromType, toType);
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/TimestampCoercer.java:128]: nit: check argument that TimestampType is not short

**Review by raunaqmorarka — APPROVED** [review]

**Pull request merged**